scripts/save_embeddings.py

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+import scanpy as sc
+import scIB
+import argparse
+import os
+import warnings
+warnings.filterwarnings('ignore')
+
+RESULT_TYPES = [
+    "full",  # reconstructed expression data
+    "embed", # embedded/latent space
+    "knn"    # corrected neighbourhood graph
+]
+
+if __name__=='__main__':
+    """
+    Save embeddings for all scenarios, methods and settings
+    """
+
+    parser = argparse.ArgumentParser(description='Save embeddings')
+
+    parser.add_argument('-i', '--input', required=True)
+    parser.add_argument('-o', '--outfile', required=True,
+                        help='Output coordinates CSV file')
+    parser.add_argument('-m', '--method', required=True, help='Name of method')
+    parser.add_argument('-r', '--result', required=True, choices=RESULT_TYPES,
+                        help='Type of result: full, embed, knn')
+    parser.add_argument('-b', '--batch_key', required=True, help='Key of batch')
+    parser.add_argument('-l', '--label_key', required=True,
+                        help='Key of annotated labels e.g. "cell_type"')
+
+    args = parser.parse_args()
+
+    outfile = args.outfile
+    method = args.method
+    result = args.result
+    setup = f'{method}_{result}'
+    batch_key = args.batch_key
+    label_key = args.label_key
+    adata = sc.read_h5ad(args.input)
+
+    print('Preparing dataset...')
+    if result == 'embed':
+        scIB.pp.reduce_data(adata, n_top_genes=None, neighbors=True,
+                            use_rep='X_emb', pca=False, umap=False)
+    elif result == 'full':
+        scIB.pp.reduce_data(adata, n_top_genes=2000, neighbors=True,
+                            use_rep='X_pca', pca=True, umap=False)
+
+    # Calculate embedding
+    if args.method.startswith('conos'):
+        print('Calculating graph embedding...')
+        sc.tl.draw_graph(adata)
+        # May be named differently depending on packages so get the last key
+        graph_key = list(adata.obsm.keys())[-1]
+        basis = graph_key.lstrip('X_')
+        label = basis.replace('draw_graph_', '').upper()
+    else:
+        print('Calculating UMAP...')
+        sc.tl.umap(adata)
+        basis = 'umap'
+        label = 'UMAP'
+
+    # Save embedding plots
+    outdir = os.path.dirname(outfile)
```

Comment:
I'd require the file as input to the script instead of matching it to the output CSV file. Then snakemake can recompute when one of the images is missing. Otherwise we could run into the issue, that csv and plot don't match, e.g. when the script fails

Reply:
Are you suggesting I have three output arguments for the script (1 for the CSV and 2 for the PNGs) or just have the output directory as an argument?

Would Snakemake be able to track things if I add extra outputs to the stage? I'm not sure what we have done for the other steps that produce multiple outputs?

Reply:
Yes, exactly.  Just as you define a named list for input you can create a named list for output and access it via `output.outputname` in the `shell` directive
```
rule abc:
...
    output:
         output1 = "umap.png"
         output2 = "other_file.txt"
    shell:
         """
         script.py { input} -o {output.output1} ... etc.
         """
```

Reply:
Ok. How's that now? I think Snakemake should be able to track the PNG files. I did try deleting one and it wasn't recreated but I think that is because it's not a direct output of `all`?

Reply:
Yeah, that's a good fix for now. It doesn't recreate the png files, because they aren't requested as input in the in the embedding_integrated and embedding_unintegrated rules. I'll  keep that as a Todo for when I restructure the pipeline, as it would be a mess to enforce it now.